benchmarks/profiling/readme.md

```diff
@@ -0,0 +1,177 @@
+# Profiling Deku-node
+
+## Profiling Deku-node using Linux [perf](https://www.brendangregg.com/perf.html)
+
+
+### Setup
+Setting `perf` to run without super-user permission (allow non-root users)
+
+```
+sudo sh -c 'echo -1 >/proc/sys/kernel/perf_event_paranoid'
+sudo sysctl -w kernel.perf_event_paranoid=-1
+sudo sh -c 'echo kernel.perf_event_paranoid=-1 > /etc/sysctl.d/local.conf'
+```
+
+`perf_event_paranoid` setting is 1:
+- `-1`: Allow use of (almost) all events by all users
+      Ignore mlock limit after perf_event_mlock_kb without CAP_IPC_LOCK
+- `>= 0`: Disallow raw and ftrace function tracepoint - access
+- `>= 1`: Disallow CPU event access
+- `>= 2`: Disallow kernel profiling
+
+To make the adjusted `perf_event_paranoid` setting permanent preserve it
+in `/etc/sysctl.conf` (e.g. `kernel.perf_event_paranoid = <setting>`)
+
+### Run
+
+```
+./profile_deku.sh run
+```
+
+Profile deku node with perf:
+- Tear down the deku node
+- Setup deku environments
+- Start deku clusters and use `perf` to get the profiling data
+
+  The samples are saved in a `perf.data` file. This does involve CPU, file system, and disk overheads to save the samples to the file system for later processing.
+ - Sleep for 10 seconds
+
+<!------------------------------------------------------------------------>
+
+##  Read `perf.data`
+### [FlameGraph](https://github.com/brendangregg/FlameGraph)
+
+Read `perf.data` using FlameGraph. FlameGraph needs to be downloaded:
+ `git clone https://github.com/brendangregg/FlameGraph`
+
+Run this script only when `perf record` is **properly terminated** (a.k.a: `./profile_deku.sh run`).
+
+```
+./profile_deku.sh flame-graph
+```
+
+Wait for awhile to let the graph generated. The result in picture as `perf-deku-flamegraph.svg`. You can use internet browser to see it `firefox perf-deku-flamegraph.svg` or chrome, etc. 
+
+This is the generated flame-graph of deku-node running with 1012 blocks:
+[flame-graph.svg](https://github.com/lykimq/deku_docs/blob/main/benchmarking/graphs/perf-deku-flamegraph.svg)
+
+
+#### Understand Flamegraph
+It shows: 
+- x-axis: show the stack profile (the sample) population, sorted alphabetically
+- y-axis: stack depth, counting from zero at the bottom.
+- Each function (stack frame) is drawn as a rectangle, with the width relative to the number of samples. The wider a frame is, the more often it was present in the stacks.
+- The top edge shows what is on-CPU.
+- The beneath it is its ancestry.
```

Comment:
I think I understand: x-axis is the aggregated time spent in each function; as such, we don't know how often we called the function. Might be that it was inefficient, might be that it's just the mainloop (but I reckon that it is already enough info to know which part of the code which should optimize).